radu.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"math/bits"
+	"os"
+	"runtime"
+	"sort"
+	"sync"
+	"unsafe"
+)
+
+const chunkSize = 1024 * 1024
+
+var ch chan fileRange
+var file *os.File
+var wg sync.WaitGroup
+
+type fileRange struct {
+	ofs, len int64
+}
+
+type partial struct {
+	ofs   int64
+	start bool
+	str   string
+}
+
+type worker struct {
+	h        hashTable
+	partials []partial
+}
+
+func (w *worker) run() {
+	buf := make([]byte, chunkSize)
+	partials := make([]partial, 0, 100)
+
+	for r := range ch {
+		b := buf[0:r.len]
+		_, err := file.ReadAt(b, r.ofs)
+		checkErr(err)
+		firstLineEnd := bytes.IndexByte(b, '\n')
+		partials = append(partials, partial{
+			ofs:   r.ofs,
+			start: false,
+			str:   string(b[:firstLineEnd+1]),
+		})
+		lastLineEnd := bytes.LastIndexByte(b, '\n')
+		if lastLineEnd < len(b)-1 {
+			partials = append(partials, partial{
+				ofs:   r.ofs + r.len,
+				start: true,
+				str:   string(b[lastLineEnd+1:]),
+			})
+		}
+		b0 := uintptr(unsafe.Pointer(&b[0]))
+		ptr := b0 + uintptr(firstLineEnd) + 1
+		end := b0 + uintptr(lastLineEnd) + 1
+		for ptr < end {
+			hash, val, nameLen, lineLen := parse(ptr)
+
+			ok, e := w.h.find(hash)
+			if ok {
+				e.count++
+				e.sum += val
+				e.minVal = min(e.minVal, val)
+				e.maxVal = max(e.maxVal, val)
+			} else {
+				e.hash = hash
+				e.name = string(b[ptr-b0 : ptr-b0+nameLen])
+				e.count = 1
+				e.sum = val
+				e.minVal = val
+				e.maxVal = val
+			}
+			ptr += lineLen
+		}
+	}
+	w.partials = partials
+	wg.Done()
+}
+
+func parse(ptr uintptr) (hash uint64, val int64, nameLen, lineLen uintptr) {
+	sep := ptr + 1
```

Comment:
Conversion from `uintptr` to `unsafe.Pointer` is not considered valid, according to docs: https://pkg.go.dev/unsafe#Pointer.

The other way around is ok though. So `ptr` should either be an `unsafe.Pointer`, or a pointer to a concrete type that you would then convert to any type through an `unsafe.Pointer` conversion.

Reply:
I think it got something to do with the fact that Go can change addresses under the hood.

Reply:
This is all in service of that idea that it may at some point move objects but it doesn't do that today (and I would venture a guess that a lot of existing unsafe code would break if that is ever implemented, including a lot of cockroach code which does these kind of conversions).

Reply:
Huh, this is all lies then? Good to know.

Reply:
But how about the stack growth case? Doesn't it change all the pointers to stack objects?

Reply:
I believe that the use of `unsafe.Pointer` causes things to escape so it would only ever refer to things on the heap. Not sure. Either way we're not dealing with a stack object in this case.